### ISSUE TITLE
Reactor stalled 264 during high write loads using cassandra-spark-connector

### ISSUE THREAD

**Issue by linux-wizard** [scylladb/scylla#7302]
*Installation details*
Scylla version (or git commit hash): scylla-4.0.0-0.20200505.d95aa77b62.x86_64
Cluster size: 3 nodes
OS (RHEL/CentOS/Ubuntu/AWS AMI): CentOS Linux release 7.7.1908 (Core)

*Hardware details (for performance issues)* 
Platform (physical/VM/cloud instance type/docker): physical
Hardware: Intel(R) Xeon(R) Gold 6230 CPU @ 2.10GHz sockets=2 cores=80 hyperthreading=enabled memory=386GB
Disks: 3 * SSD

During high write load period (~80K INSERT/s per node, using cassandra-spark-connector and batches), we have from time to time some reactor stalled messages. Most fo them are around 264ms

```
[root@xx-xxxxx-xxxx ~]# journalctl --since "2 days ago" -u scylla-server -o cat | grep "Reactor stalled for"
Reactor stalled for 263 ms on shard 49.
Reactor stalled for 263 ms on shard 18.
Reactor stalled for 264 ms on shard 6.
Reactor stalled for 263 ms on shard 23.
Reactor stalled for 263 ms on shard 48.
Reactor stalled for 264 ms on shard 6.
Reactor stalled for 263 ms on shard 46.
Reactor stalled for 263 ms on shard 8.
Reactor stalled for 263 ms on shard 57.
Reactor stalled for 262 ms on shard 57.
Reactor stalled for 263 ms on shard 54.
Reactor stalled for 263 ms on shard 12.
Reactor stalled for 263 ms on shard 52.
Reactor stalled for 262 ms on shard 73.
Reactor stalled for 264 ms on shard 6.
Reactor stalled for 263 ms on shard 14.
Reactor stalled for 262 ms on shard 19.
Reactor stalled for 262 ms on shard 55.
Reactor stalled for 262 ms on shard 76.
Reactor stalled for 262 ms on shard 73.
Reactor stalled for 262 ms on shard 73.
Reactor stalled for 263 ms on shard 2.
Reactor stalled for 263 ms on shard 49.
Reactor stalled for 264 ms on shard 69.
Reactor stalled for 264 ms on shard 8.
Reactor stalled for 263 ms on shard 25.
Reactor stalled for 262 ms on shard 42.
Reactor stalled for 263 ms on shard 27.
Reactor stalled for 263 ms on shard 36.
Reactor stalled for 263 ms on shard 77.
```

### Backtrace

```
Sep 25 06:49:54 scylla[41350]: Reactor stalled for 264 ms on shard 7.
Sep 25 06:49:54 scylla[41350]: Backtrace:
Sep 25 06:49:54 scylla[41350]: 0x0000000002ed8bc2
Sep 25 06:49:54 scylla[41350]: 0x0000000002e7e73b
Sep 25 06:49:54 scylla[41350]: 0x0000000002e7ec62
Sep 25 06:49:54 scylla[41350]: 0x0000000002e7ef97
Sep 25 06:49:54 scylla[41350]: 0x00007f6565f36b1f
Sep 25 06:49:54 scylla[41350]: 0x00000000029396f6
Sep 25 06:49:54 scylla[41350]: 0x000000000289e2e5
Sep 25 06:49:54 scylla[41350]: 0x00000000028ad99f
Sep 25 06:49:54 scylla[41350]: 0x0000000002e7ac30
Sep 25 06:49:54 scylla[41350]: 0x0000000002e7ae3f
Sep 25 06:49:54 scylla[41350]: 0x0000000002eb2df5
Sep 25 06:49:54 scylla[41350]: 0x0000000002ec141c
Sep 25 06:49:54 scylla[41350]: 0x0000000002e4caed
Sep 25 06:49:54 scylla[41350]: /opt/scylladb/libreloc/libpthread.so.0+0x00000000000094e1
Sep 25 06:49:54 scylla[41350]: /opt/scylladb/libreloc/libc.so.6+0x00000000001016a2
```

**Comment by linux-wizard** [comment]
decoded backtrace

```
addr2line -fCipe /usr/lib/debug/opt/scylladb/libexec/scylla-4.0.0-0.20200505.d95aa77b62.x86_64.debug  0x0000000002ed8bc2 0x0000000002e7e73b 0x0000000002e7ec62 0x0000000002e7ef97 0x00007f6565f36b1f 0x00000000029396f6 0x000000000289e2e5 0x00000000028ad99f 0x0000000002e7ac30 0x0000000002e7ae3f 0x0000000002eb2df5 0x0000000002ec141c 0x0000000002e4caed /opt/scylladb/libreloc/libpthread.so.0+0x00000000000094e1 /opt/scylladb/libreloc/libc.so.6+0x00000000001016a2

void seastar::backtrace<seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}>(seastar::backtrace_buffer::append_backtrace()::{lambda(seastar::frame)#1}&&) at /usr/include/fmt/format.h:1158
seastar::backtrace_buffer::append_backtrace() at /usr/include/fmt/format.h:1158
 (inlined by) print_with_backtrace at ./build/release/seastar/./seastar/src/core/reactor.cc:755
seastar::internal::cpu_stall_detector::generate_trace() at /usr/include/fmt/format.h:1158
seastar::internal::cpu_stall_detector::maybe_report() at /usr/include/fmt/format.h:1158
 (inlined by) seastar::internal::cpu_stall_detector::on_signal() at ./build/release/seastar/./seastar/src/core/reactor.cc:1062
 (inlined by) seastar::reactor::block_notifier(int) at ./build/release/seastar/./seastar/src/core/reactor.cc:1185
?? ??:0
boost::intrusive::default_rbtree_node_traits_impl<void*>::get_left(boost::intrusive::rbtree_node<void*>* const&) at ././seastar/include/seastar/core/future.hh:1675
 (inlined by) boost::intrusive::bstree_algorithms_base<boost::intrusive::rbtree_node_traits<void*, false> >::minimum(boost::intrusive::rbtree_node<void*>*) at /usr/include/boost/intrusive/detail/bstree_algorithms_base.hpp:100
 (inlined by) boost::intrusive::bstree_algorithms_base<boost::intrusive::rbtree_node_traits<void*, false> >::next_node(boost::intrusive::rbtree_node<void*>* const&) at /usr/include/boost/intrusive/detail/bstree_algorithms_base.hpp:49
 (inlined by) boost::intrusive::bstree_algorithms<boost::intrusive::rbtree_node_traits<void*, false> >::size(boost::intrusive::rbtree_node<void*> const* const&) at /usr/include/boost/intrusive/bstree_algorithms.hpp:674
 (inlined by) boost::intrusive::bstree_impl<boost::intrusive::mhtraits<cache_entry, boost::intrusive::set_member_hook<boost::intrusive::link_mode<(boost::intrusive::link_mode_type)2>, void, void, void>, &cache_entry::_cache_link>, void, cache_entry::compare, unsigned long, false, (boost::intrusive::algo_types)5, void>::size() const at /usr/include/boost/intrusive/bstree.hpp:966
 (inlined by) row_cache::partitions() const at ././row_cache.hh:561
 (inlined by) operator() at ./api/cache_service.cc:255
 (inlined by) operator() at ./api/column_family.hh:91
 (inlined by) _M_invoke at /usr/include/c++/9/bits/std_function.h:286
std::function<std::unique_ptr<std::any, std::default_delete<std::any> > (table&)>::operator()(table&) const at /usr/include/fmt/format.h:1158
 (inlined by) api::map_reduce_column_families_locally::operator()(database&) const::{lambda(std::pair<utils::UUID, seastar::lw_shared_ptr<table> > const&)#1}::operator()(std::pair<utils::UUID, seastar::lw_shared_ptr<table> > const&) const at ./api/column_family.hh:78
 (inlined by) seastar::future<> seastar::do_void_futurize_helper<void>::apply<api::map_reduce_column_families_locally::operator()(database&) const::{lambda(std::pair<utils::UUID, seastar::lw_shared_ptr<table> > const&)#1}&, std::pair<utils::UUID const, seastar::lw_shared_ptr<table> >&>(api::map_reduce_column_families_locally::operator()(database&) const::{lambda(std::pair<utils::UUID, seastar::lw_shared_ptr<table> > const&)#1}&, std::pair<utils::UUID const, seastar::lw_shared_ptr<table> >&) at ././seastar/include/seastar/core/future.hh:1608
 (inlined by) seastar::future<> seastar::futurize<void>::apply<api::map_reduce_column_families_locally::operator()(database&) const::{lambda(std::pair<utils::UUID, seastar::lw_shared_ptr<table> > const&)#1}&, std::pair<utils::UUID const, seastar::lw_shared_ptr<table> >&>(api::map_reduce_column_families_locally::operator()(database&) const::{lambda(std::pair<utils::UUID, seastar::lw_shared_ptr<table> > const&)#1}&, std::pair<utils::UUID const, seastar::lw_shared_ptr<table> >&) at ././seastar/include/seastar/core/future.hh:1657
 (inlined by) seastar::future<> seastar::do_for_each<std::__detail::_Node_iterator<std::pair<utils::UUID const, seastar::lw_shared_ptr<table> >, false, true>, api::map_reduce_column_families_locally::operator()(database&) const::{lambda(std::pair<utils::UUID, seastar::lw_shared_ptr<table> > const&)#1}>(std::__detail::_Node_iterator<std::pair<utils::UUID const, seastar::lw_shared_ptr<table> >, false, true>, seastar::future<>, api::map_reduce_column_families_locally::operator()(database&) const::{lambda(std::pair<utils::UUID, seastar::lw_shared_ptr<table> > const&)#1}) at ././seastar/include/seastar/core/future-util.hh:607
 (inlined by) seastar::future<> seastar::do_for_each<std::unordered_map<utils::UUID, seastar::lw_shared_ptr<table>, std::hash<utils::UUID>, std::equal_to<utils::UUID>, std::allocator<std::pair<utils::UUID const, seastar::lw_shared_ptr<table> > > >, api::map_reduce_column_families_locally::operator()(database&) const::{lambda(std::pair<utils::UUID, seastar::lw_shared_ptr<table> > const&)#1}>(std::unordered_map<utils::UUID, seastar::lw_shared_ptr<table>, std::hash<utils::UUID>, std::equal_to<utils::UUID>, std::allocator<std::pair<utils::UUID const, seastar::lw_shared_ptr<table> > > >&, api::map_reduce_column_families_locally::operator()(database&) const::{lambda(std::pair<utils::UUID, seastar::lw_shared_ptr<table> > const&)#1}) at ././seastar/include/seastar/core/future-util.hh:637
 (inlined by) api::map_reduce_column_families_locally::operator()(database&) const at ./api/column_family.hh:77
seastar::future<std::unique_ptr<std::any, std::default_delete<std::any> > > seastar::sharded<database>::map_reduce0<api::map_reduce_column_families_locally, std::unique_ptr<std::any, std::default_delete<std::any> >, std::function<std::unique_ptr<std::any, std::default_delete<std::any> > (std::unique_ptr<std::any, std::default_delete<std::any> >, std::unique_ptr<std::any, std::default_delete<std::any> >)> >(api::map_reduce_column_families_locally, std::unique_ptr<std::any, std::default_delete<std::any> >, std::function<std::unique_ptr<std::any, std::default_delete<std::any> > (std::unique_ptr<std::any, std::default_delete<std::any> >, std::unique_ptr<std::any, std::default_delete<std::any> >)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}::operator()() const at /usr/include/fmt/format.h:1158
 (inlined by) seastar::future<std::unique_ptr<std::any, std::default_delete<std::any> > > seastar::futurize<seastar::future<std::unique_ptr<std::any, std::default_delete<std::any> > > >::apply<seastar::future<std::unique_ptr<std::any, std::default_delete<std::any> > > seastar::sharded<database>::map_reduce0<api::map_reduce_column_families_locally, std::unique_ptr<std::any, std::default_delete<std::any> >, std::function<std::unique_ptr<std::any, std::default_delete<std::any> > (std::unique_ptr<std::any, std::default_delete<std::any> >, std::unique_ptr<std::any, std::default_delete<std::any> >)> >(api::map_reduce_column_families_locally, std::unique_ptr<std::any, std::default_delete<std::any> >, std::function<std::unique_ptr<std::any, std::default_delete<std::any> > (std::unique_ptr<std::any, std::default_delete<std::any> >, std::unique_ptr<std::any, std::default_delete<std::any> >)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}&>(seastar::future<std::unique_ptr<std::any, std::default_delete<std::any> > > seastar::sharded<database>::map_reduce0<api::map_reduce_column_families_locally, std::unique_ptr<std::any, std::default_delete<std::any> >, std::function<std::unique_ptr<std::any, std::default_delete<std::any> > (std::unique_ptr<std::any, std::default_delete<std::any> >, std::unique_ptr<std::any, std::default_delete<std::any> >)> >(api::map_reduce_column_families_locally, std::unique_ptr<std::any, std::default_delete<std::any> >, std::function<std::unique_ptr<std::any, std::default_delete<std::any> > (std::unique_ptr<std::any, std::default_delete<std::any> >, std::unique_ptr<std::any, std::default_delete<std::any> >)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}&) at ././seastar/include/seastar/core/future.hh:1674
 (inlined by) seastar::smp_message_queue::async_work_item<seastar::future<std::unique_ptr<std::any, std::default_delete<std::any> > > seastar::sharded<database>::map_reduce0<api::map_reduce_column_families_locally, std::unique_ptr<std::any, std::default_delete<std::any> >, std::function<std::unique_ptr<std::any, std::default_delete<std::any> > (std::unique_ptr<std::any, std::default_delete<std::any> >, std::unique_ptr<std::any, std::default_delete<std::any> >)> >(api::map_reduce_column_families_locally, std::unique_ptr<std::any, std::default_delete<std::any> >, std::function<std::unique_ptr<std::any, std::default_delete<std::any> > (std::unique_ptr<std::any, std::default_delete<std::any> >, std::unique_ptr<std::any, std::default_delete<std::any> >)>)::{lambda(unsigned int)#1}::operator()(unsigned int) const::{lambda()#1}>::run_and_dispose() at ././seastar/include/seastar/core/smp.hh:215
seastar::reactor::run_tasks(seastar::reactor::task_queue&) at /usr/include/fmt/format.h:1158
seastar::reactor::run_some_tasks() at /usr/include/fmt/format.h:1158
seastar::reactor::run_some_tasks() at /usr/include/fmt/format.h:1158
 (inlined by) seastar::reactor::run() at ./build/release/seastar/./seastar/src/core/reactor.cc:2693
seastar::smp::configure(boost::program_options::variables_map, seastar::reactor_config)::{lambda()#3}::operator()() const at /usr/include/fmt/format.h:1158
std::function<void ()>::operator()() const at /usr/include/c++/9/bits/std_function.h:688
 (inlined by) seastar::posix_thread::start_routine(void*) at ./build/release/seastar/./seastar/src/core/posix.cc:52
?? ??:0
?? ??:0
```





--

**Comment by avikivity** [comment]
In 4.0 row_cache::_partitions looks like this:

```c++
    using partitions_type = bi::set<cache_entry,
        bi::member_hook<cache_entry, cache_entry::cache_link_type, &cache_entry::_cache_link>,
        bi::constant_time_size<false>, // we need this to have bi::auto_unlink on hooks
        bi::compare<cache_entry::compare>>;
```

this is an API call to compute the number of partitions in cache (no idea why spark wants that, but maybe it's part of a larger nodetool call), and we have `constant_time_size` disabled.

**Comment by avikivity** [comment]
Was fixed in 83fe0427d29e6474b900e5324f1fc015557ca3f0 (scylla 4.1).
